Add resource type sqlServerInstances

caod opened the pull request:
* add to resource provider - Microsoft.AzureData

* exactly same REST API definitions as existing resource types of the
  same resource provider

### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [x] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [x] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [x] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

Commit:
Add resource type sqlServerInstances

* add to resource provider - Microsoft.AzureData

* exactly same REST API definitions as existing resource types of the
  same resource provider

AutorestCI commented:
You don't have permission to trigger SDK Automation.
Please add yourself to ***Azure*** group from opensource portal if you are MSFT employee,
or please ask reviewer to add comment *** /openapibot sdkautomation ***.
Please ask tih@microsoft.com (or NullMDR in github) for additional help.

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
/azp run automation - sdk

caod commented:
The two failed checks "automation - sdk" and "automation - sdk (SDK Go)" are existing baseline failures even without changes from this PR.

caod commented:
The link of "OpenAPI Hub" (https://portal.azure-devex-tools.com/github-connect?returnTo=%2F) shows an empty page to me. Do I need to request any access permissions?

pilor commented on specification/azuredata/resource-manager/Microsoft.AzureData/preview/2019-07-24-preview/azuredata.json:
```diff
@@ -1984,6 +2272,88 @@
+      "properties": {
+        "version": {
+          "type": "string",
+          "description": "null"
```
>"null" [](start = 25, length = 6)

All properties are missing descriptions

caod replied:
I saw existing properties in this file has description set as null. It is ok to leave description as null for the new resource type too?

pilor replied:
The descriptions are used directly in the generated REST documentation and SDKs.  They should be human readable and descriptive.  They also make these reviews a lot easier

caod replied:
Thanks for the explanation. I will add descriptions.

pilor commented on specification/azuredata/resource-manager/Microsoft.AzureData/preview/2019-07-24-preview/azuredata.json:
```diff
@@ -1984,6 +2272,88 @@
+      "description": "Properties of SqlServerInstance.",
+      "type": "object",
+      "properties": {
+        "version": {
```
>version" [](start = 9, length = 8)

what properties are required on a create?  Are any properties read only?  Please mark them as such

caod replied:
All properties are needed for creating a new resource of this resource type. If later any of them can be optional, will update the spec.

pilor replied:
They should be marked as required then

caod replied:
Will mark them as "required"

pilor replied:
Sorry, required is a string array at the level above the properties.  See how "name" is marked required in the "Sku" definition in this file as an example

pilor commented on specification/azuredata/resource-manager/Microsoft.AzureData/preview/2019-07-24-preview/azuredata.json:
```diff
@@ -1984,6 +2272,88 @@
+        "updateTime": {
+          "type": "string",
+          "description": "null"
+        },
```
these should be part of the top-level systemData contract for new resources.  Details here:  [https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources)

caod replied:
This new resource type shared the common code with existing resource types of this resource provider. systemData contract, as a 2020 new requirement, is not added into the common infra of this resource provider yet. Consider this new resource type is for private preview purpose, is it ok to be an exception? Otherwise, all published existing resource types of this resource provider share the same issue too.

pilor replied:
Its not a mandate yet so yes, its technically OK.  You should aim to have it implemented before you leave private preview

pilor commented on specification/azuredata/resource-manager/Microsoft.AzureData/preview/2019-07-24-preview/azuredata.json:
```diff
@@ -1984,6 +2272,88 @@
+          "type": "string",
+          "description": "null"
+        },
+        "status": {
```
>"status" [](start = 8, length = 8)

what is status used for?  Is this a provisioningState?

caod replied:
status can be "Connected" or "Disconnected". It is a Azure Arc resource type.

pilor commented on specification/azuredata/resource-manager/Microsoft.AzureData/preview/2019-07-24-preview/azuredata.json:
```diff
@@ -1984,6 +2272,88 @@
+          "type": "string",
+          "description": "null"
+        },
+        "containerResourceId": {
```
>containerResourceId [](start = 9, length = 19)

what is a container in this context?

caod replied:
It is resource id of Azure Arc for Servers. Only an on-premise machine is registered for Azure Arc for Servers can register its SQL Server instance as a sqlServerInstances resource.

pilor commented on specification/azuredata/resource-manager/Microsoft.AzureData/preview/2019-07-24-preview/azuredata.json:
```diff
@@ -1984,6 +2272,88 @@
+          "type": "string",
+          "description": "null"
+        },
+        "edition": {
```
>"edition": { [](start = 8, length = 12)

can version and edition be string enums?

caod replied:
It can be string enums. Currently all existing resource types of thie resource provider define its property type as string. To support other types of property, an infra update is needed. Considering this resource type is for private preview, can we ask for an exception to keep it as string for now?

pilor reviewed (changes requested):
:clock1:

Commit:
Add property description and mark required properties
  # update examples to include required properties

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

pilor commented on specification/azuredata/resource-manager/Microsoft.AzureData/preview/2019-07-24-preview/azuredata.json:
```diff
@@ -1984,6 +2272,97 @@
+        "edition",
+        "containerResourceId",
+        "createTime",
+        "updateTime",
```
createTime, updateTime, and status aren't actually required to be entered by the user right?  Wouldn't they be readonly, generated by the RP?

caod replied:
"status" needs to be input by user because we are not sure if an auto detection will be added at this moment. But we may re-consider it later.
For "createTime" and "updateTime", you are right that it should be auto generated. The code is not in place yet for the private preview timeline.

pilor replied:
createTime and updateTime should still be removed from the required list and marked readonly.  The user should not b entering these even if you aren't populating them in the backend yet

caod replied:
will remove createTime and updateTime from the required list.

caod replied:
createTime and updateTime are set to readonly now.

AutorestCI commented:
/azp run automation - sdk

Commit:
set createTime and updateTime property to readonly

  update CreateOrUpdateSqlServerInstance.json sample to not include readonly
  properties in its request.

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

AutorestCI commented:
/azp run automation - sdk

lirenhe commented:
/azp run automation - sdk
